devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,378 @@
+            # There will most likely be more than one MatVecAction
+            # associated with the target e.g interior matvec + BC matvecs
+            if matvec.expr.rhs.target != target:
+                continue
+            solver = generate_solver_calls(solver_objs, objs, matvec, target)
+            setup.extend(solver)
+            break
+
+        # Create the body of the matrix-vector callback for target
+        for iter, (matvec,) in matvec_mapper.items():
+            if matvec.expr.rhs.target != target:
+                continue
+            matvec_body = matvec_body_list._rebuild(
+                body=[matvec_body_list.body, iter[0]])
+            matvec_body_list = matvec_body_list._rebuild(body=matvec_body)
+            # Remove the iteration loop from the main kernel encapsulating
+            # the matvec equations since they are moved to the callback
+            subs.update({iter[0]: None})
+
+        # Create the matvec callback and operation for each target
+        matvec_callback, matvec_op = create_matvec_callback(
+            target, matvec_body_list, solver_objs, objs,
+            struct)
+
+        setup.extend([matvec_op, BlankLine])
+        efuncs[matvec_callback.name] = matvec_callback
+
+    # Remove the LinSolveExpr's from iet and efuncs
+    subs.update(rebuild_expr_mapper(iet))
+    iet = Transformer(subs).visit(iet)
+    efuncs = transform_efuncs(efuncs, struct)
+
+    body = iet.body._rebuild(init=init, body=core + tuple(setup) + iet.body.body)
+    iet = iet._rebuild(body=body)
+
+    metadata = core_metadata()
+    metadata.update({'efuncs': tuple(efuncs.values())})
+
+    return iet, metadata
+
+
+def init_petsc(**kwargs):
+
+    # Initialize PETSc -> for now, assuming all solver options have to be
+    # specifed via the parameters dict in PETScSolve
+    # TODO: Are users going to be able to use PETSc command line arguments?
+    # In firedrake, they have an options_prefix for each solver, enabling the use
+    # of command line options
+    initialize = petsc_call('PetscInitialize', [Null, Null, Null, Null])
+
+    return tuple([petsc_func_begin_user, initialize])
+
+
+def build_petsc_struct(iet):
+    # Place all context data required by the shell routines
+    # into a PETScStruct
+    usr_ctx = []
+    basics = FindSymbols('basics').visit(iet)
+    avoid = FindSymbols('dimensions|indexedbases').visit(iet)
+    usr_ctx.extend(data for data in basics if data not in avoid)
+
+    return PETScStruct('ctx', usr_ctx)
+
+
+def make_core_petsc_calls(target, struct, objs, **kwargs):
+    # MPI
+    call_mpi = petsc_call_mpi('MPI_Comm_size', [objs['comm'], Byref(objs['size'])])
+
+    # Create DMDA
+    dmda = create_dmda(target, objs)
+    dm_setup = petsc_call('DMSetUp', [objs['da']])
+    dm_app_ctx = petsc_call('DMSetApplicationContext', [objs['da'], struct])
+    dm_mat_type = petsc_call('DMSetMatType', [objs['da'], 'MATSHELL'])
+
+    return (
```

Comment:
No need for tuple brackets. Should be `return petsc_func_begin_user, call_mpi, ...`